### PR TITLE
Lxd first time failure

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

There are two parts to this change, the first one is the removal of the old mechanism for enabling the https listener. The second part is then implementing a retry mechanism where by we request a new local provider for the lxd client so that we can re-request the connection information correctly.

> Why is this change needed?

Juju bootstrap on bionic would fail on a lxd localhost, which caused the user to manually retry. The subsequent retry would work fine. 

## QA steps

> How do we verify that the change works?

Assume you have the source code, you can do the following:

On your local shell
```shell
lxc launch ubuntu:18.04 bionic
make go-build go-install
lxc file push ~/go/bin/juju* bionic/
lxc exec bionic bash
```
With in the lxc bionic shell 
```shell
lxd init --auto
mv /juju* ~
./juju bootstrap localhost test
```

## Documentation changes

> Does it affect current user workflow? CLI? API?

Bionic first time setup with lxd should work first time, instead of requiring a retry manually.

## Bug reference

> Does this change fix a bug? Please add a link to it.

- https://trello.com/c/bU1TseDj/258-bootstrapping-lxd-on-bionic-first-time-failure
- https://bugs.launchpad.net/juju/+bug/1771887

### CODE REVIEW

File: provider/lxd/provider.go [hunk 1]
```diff
@@ -166,25 +169,53 @@ func (p *environProvider) getLocalHostAddress(ctx environs.FinalizeCloudContext)
...
+	errNotExists := errors.New("not-exists")
+	retryArgs := retry.CallArgs{
+		Clock: p.clock(),
+		IsFatalError: func(err error) bool {
+			return errors.Cause(err) != errNotExists
+		},
+		Func: func() error {
+			// Requesting a newLocalRawProvider forces a new connection, so
+			// that when we GetConnectionInfo it gets the required addresses.
+			raw, err := p.newLocalRawProvider()
```
Comment (manadart) [inline provider/lxd/provider.go]: This should just use ```ConnectLocal``` from _container.lxd_.

```newLocalRawProvider``` calls through to _tools/lxdclient_ and re-verifies the networking config, which will cause multiple log messages. We want to ditch the use of ```rawProvider``` anyway.

This should mean that you don't need to add ```EnableHTTPSListener``` to the ```Server``` interface, or in turn to the testing stub.

File: provider/lxd/provider.go [hunk 2]
```diff
@@ -166,25 +169,53 @@ func (p *environProvider) getLocalHostAddress(ctx environs.FinalizeCloudContext)
...
+			if err != nil {
+				return errors.Trace(err)
+			}
+
+			cInfo, err := raw.GetConnectionInfo()
```
Comment (manadart) [inline provider/lxd/provider.go]: Did you try ```GetServer``` here? 

It has ```Environment.Addresses``` on it. If the addresses turned up there, we wouldn't have to reconnect each time.

File: provider/lxd/provider.go [hunk 3]
```diff
@@ -153,10 +156,22 @@ func (p *environProvider) FinalizeCloud(
 }
 
 func (p *environProvider) getLocalHostAddress(ctx environs.FinalizeCloudContext) (string, error) {
-	raw, err := p.newLocalRawProvider()
+	raw, err := lxd.NewLocalServer()
```
Comment (manadart) [inline provider/lxd/provider.go]: I would name this "svr" or "server".

File: provider/lxd/provider.go [hunk 4]
```diff
@@ -166,25 +181,53 @@ func (p *environProvider) getLocalHostAddress(ctx environs.FinalizeCloudContext)
...
-	}
-	var found bool
-	for _, addr := range cInfo.Addresses {
-		if strings.HasPrefix(addr, hostAddress+":") {
-			hostAddress = addr
-			found = true
-			break
-		}
+
+	// The following retry mechanism is required for bionic, where the new
```
Comment (manadart) [inline provider/lxd/provider.go]: "required for newer LXD versions..."

File: provider/lxd/provider_test.go [hunk 5]
```diff
@@ -120,9 +124,7 @@ func (s *providerSuite) TestFinalizeCloudNotListening(c *gc.C) {
 		}},
 	})
 	c.Assert(err, gc.NotNil)
-	c.Assert(err.Error(), gc.Equals,
-		`LXD is not listening on address https://8.8.8.8 `+
-			`(reported addresses: [127.0.0.1:1234 1.2.3.4:1234])`)
+	c.Assert(err, gc.ErrorMatches, `LXD is not listening on address https:\/\/8\.8\.8\.8 \(reported addresses\: \[.+\]\)`)
 }
 
 func (s *providerSuite) TestFinalizeCloudAlreadyListeningHTTPS(c *gc.C) {
```
Comment (manadart) [inline provider/lxd/provider_test.go]: This one could be addressed, but I am OK with leaving it until we rip out ```rawProvider``` altogether.

Comment (SimonRichardson) [reply provider/lxd/provider_test.go]: Agreed

**Comment by jameinel** [comment]
[Bootstrapping LXD on Bionic - First Time Failure](https://trello.com/c/bU1TseDj/258-bootstrapping-lxd-on-bionic-first-time-failure)

**Comment by SimonRichardson** [comment]
Don't merge, it's broken on xenial...

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**